### ISSUE TITLE
`FileSystemHandle.name` of the root directory

### ISSUE THREAD

**Issue by ichaoX** [WICG/file-system-access#437]
The `FileSystemHandle.name` of the root directory may not be a [valid file name](https://fs.spec.whatwg.org/#valid-file-name) (e.g. [in bucket file system, it is an empty string](https://fs.spec.whatwg.org/#filesystemhandle:~:text=Note:%20This%20is%20a%20bit,be%20a%20valid%20file%20name.)), and developers may need to be aware of these special cases.

Is there a universal rule for obtaining the `FileSystemHandle.name` based on the root directory?

The following names are just my predictions and may differ from the current implementation of Chromium.

| path | FileSystemHandle.name |
|----|----|
| '/' | '' |
| 'D:\\' | 'D:' |
| '\\\\smb.example.com\\share\\' | 'share' |
| 'http://webdav.example.com/' | 'webdav.example.com' |

**Comment by a-sully** [comment]
The root directory of a Bucket File System is a special case. It is always the empty string; the only directory for which the `name` is not a "valid file name".

For all other `FileSystemHandle`s (including other "root" directories), the `name` is specified by https://fs.spec.whatwg.org/#locating-an-entry:

> entry’s [name](https://fs.spec.whatwg.org/#entry-name) is the last [item](https://infra.spec.whatwg.org/#list-item) of locator’s [path](https://fs.spec.whatwg.org/#locator-path).

**Comment by ichaoX** [comment]
@a-sully Thanks for your reply.
> entry’s [name](https://fs.spec.whatwg.org/#entry-name) is the last [item](https://infra.spec.whatwg.org/#list-item) of locator’s [path](https://fs.spec.whatwg.org/#locator-path).

> A file system path is a [list](https://infra.spec.whatwg.org/#list) of one or more [strings](https://infra.spec.whatwg.org/#string). This may be a virtual path that is mapped to real location on disk or in memory, may correspond directly to a path on the local file system, or may not correspond to any file on disk at all. The actual physical location of the corresponding [file system entry](https://fs.spec.whatwg.org/#entry) is [implementation-defined](https://infra.spec.whatwg.org/#implementation-defined).

These descriptions are a bit unclear to me, does it mean that different User-Agents may return different `FileSystemHandle.name` for the same path 'D:\\'? (e.g. it's `'\'` in Chromium, but `'D:'` in [Firefox Extension](https://github.com/ichaoX/ext-file).)

As far as I know if `FileSystemHandle.name` is `'\'` or `'D:'` both will make [vscode.dev](https://vscode.dev) unavailable.

**Comment by a-sully** [comment]
Ah I see. This is currently not well-specified. The picker methods say to create:

> a new [FileSystemFileHandle](https://fs.spec.whatwg.org/#filesystemfilehandle) associated with _entry_

... which has a few problems:

- This language doesn't make sense anymore in the wake of https://github.com/whatwg/fs/pull/96. This spec needs to be updated accordingly (related to #408, though realistically this is blocked on https://github.com/whatwg/fs/pull/131)
- The properties of the `entry` yielded from the file picker are not very well-defined:

  > "Let _entry_ be a [file entry](https://fs.spec.whatwg.org/#file) representing the selected file"

  This is very hand-wavy. We may want to make this more precise, perhaps poaching ideas from [the `<input type="file">` spec](https://html.spec.whatwg.org/multipage/input.html#file-upload-state-(type=file)). Would you like to either file a new issue or update the title+description of this issue to reflect that we need to better specify the `name` of the `FileSystemHandle` _selected from a picker_?

All that being said, `'\'` is not a valid file name (nor would it be once we [update the criteria for a valid file name](https://github.com/whatwg/fs/issues/93#issuecomment-1402866283)). Chromium appears to strip the drive letter from the path if you select a root (though we [add it back in UI surfaces](https://crsrc.org/c/chrome/browser/ui/views/file_system_access/file_system_access_ui_helpers.cc?q=f:file_system_access%20-f:gen%20%22drive%20letter%22&ss=chromium), e.g. for permissions) though I'm not sure off the top of my head why (@mkruisselbrink may know). Would you mind filing a [new Chrome bug](https://new.crbug.com/) and pasting the URL here?

> [Firefox Extension](https://github.com/ichaoX/ext-file)

This looks very cool, by the way! :)

**Comment by ichaoX** [comment]
The path "/" likely doesn't have a [valid file name](https://fs.spec.whatwg.org/#valid-file-name) either.

I cannot confirm this behavior with Chromium at the moment, but technically it should be able to be open by `DataTransferItem.getAsFileSystemHandle()`.

**Comment by a-sully** [comment]
> it's `'\'` in Chromium, but `'D:'` in [Firefox Extension](https://github.com/ichaoX/ext-file)

I agree that both `'\'` and `'/'` are not valid names - or at least won't be once we address https://github.com/whatwg/fs/issues/93

That issue relates to paths for the Bucket File System, though. What's not specified - but is necessary to safely allow access to the local file system - is the _additional_ restrictions the user agent may put on file names. I imagine these restrictions will be non-normative once they are added to the spec.

For example, it looks like [Chromium does not allow the colon character in file names](https://crsrc.org/c/base/i18n/file_util_icu.cc;drc=00e65cdb5f298e45953893a134c932643346e121;l=124), which explains why the drive letter is stripped....

> If an invalid file name is a bug, then what should the name be?

...Good question! :)

The least-bad option might be to carve out an exception for drive roots ('/', 'D:\', etc) in the same way we have done for the Bucket File System root. Whether that's `'/'` or `'\'` or `''` or something else, I'm indifferent to... But it seems like it would be nice to pick one and specify it

**Comment by ichaoX** [comment]
I think that if the user agent cannot generate a valid file name for the root directory, reusing the empty string of the bucket file system is better than using path separators. Because developers only need to additionally consider this one special case.

However, probably generating a  [valid file name](https://fs.spec.whatwg.org/#valid-file-name) (e.g. `"[root]"`) would have better compatibility than using an empty string.
